### ISSUE TITLE
Compaction in SizeTiered it is not clear how it decides what will be the next compaction

### ISSUE THREAD

**Issue by roydahan** [scylladb/scylla#2095]
*Installation details*
Scylla version (or git commit hash): 1.6+

we know there is no compaction level 3 of ~900M files ... why does compaction select files in level 1 - why does it not start a compaction in level 3 when we have 3 such files ?

```
2017-02-19 10:19:47,453 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 0 (generation,data_size) (29029,13981415) (29673,14013748)
2017-02-19 10:19:47,454 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 1 (generation,data_size) (30163,225065841) (30219,225071056) (30205,225073142) (30191,225073142)
2017-02-19 10:19:47,454 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 2 (generation,data_size) (28469,496042456) (29001,502129404) (29645,504166383)
2017-02-19 10:19:47,456 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 3 (generation,data_size) (29883,900272751) (29967,900273794) (29785,900279009) (30135,900288396) (30037,900300912) (29715,1
350419035)
2017-02-19 10:19:47,456 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 4 (generation,data_size) (29827,3601140025)
2017-02-19 10:19:47,457 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 5 (generation,data_size) (29379,7799348529) (27237,7900377681) (25375,7908287793) (26887,7930037472)
2017-02-19 10:19:47,457 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 6 (generation,data_size) (23863,13389822271) (27601,14450472960) (25417,20730661742)
2017-02-19 10:19:47,457 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 7 (generation,data_size) (17241,26161382303) (6811,28620821152)
2017-02-19 10:19:47,459 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 err    |  [shard 7] compaction - size_tiered: level 8 (generation,data_size) (22043,41928297530) (14063,45557723715)
2017-02-19 10:19:47,744 remote           L0194 DEBUG| [52.91.169.183] [stdout] Feb 19 10:19:47 info   |  [shard 7] compaction - Compacting [/var/lib/scylla/data/keyspace1/standard1-337cbc00f45c11e68098000000000002/keyspace1-standard1-ka-30163-Data.db:level=0, /var/lib/s
cylla/data/keyspace1/standard1-337cbc00f45c11e68098000000000002/keyspace1-standard1-ka-30219-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1-337cbc00f45c11e68098000000000002/keyspace1-standard1-ka-30205-Data.db:level=0, /var/lib/scylla/data/keyspace1/standard1
-337cbc00f45c11e68098000000000002/keyspace1-standard1-ka-30191-Data.db:level=0, ]
```

**Comment by nyh** [comment]
@raphaelsc - I already commented in the past that it is confusing to use the word "level" in the context of size-tiered compaction. In addition to the leveled-compaction meaning (irrelevant here), there are probably two additional possible meanings: one is size-tiered-compaction notion of "sstable buckets", and one is compaction-manager's notion of "weight classes" (sizes, actually) for parallel compaction. I don't know what the levels in the printouts you see actually refer to because I can't find the code that prints them (?!).

In any case, if I understand correctly, when we already have a small compaction going on, we do NOT start another large compaction in parallel. The thinking is that we know that the small compaction will finish quickly, and the large compaction will not be significantly hurt by waiting a little more (the large compaction itself will take much longer than the small compaction we're waiting for). Moreover, the primary benefit of compaction is reducing the number of sstables, and small compactions reduce this number the quickest.
So we only start a new parallel compaction if we are in the middle of a large compaction and now want to start a small one.

As to why we choose to start the small compaction in the first place and not the large one - I don't remember ( @raphaelsc probably does) but I'm guessing it's for the same reasons I mentioned above).

**Comment by slivne** [comment]
the printouts are here

```
diff --git a/sstables/compaction_strategy.cc b/sstables/compaction_strategy.cc
index 8d80dd4..0094726 100644
--- a/sstables/compaction_strategy.cc
+++ b/sstables/compaction_strategy.cc
@@ -609,6 +609,17 @@ compaction_descriptor size_tiered_compaction_strategy::get_sstables_for_compacti
 
     auto buckets = get_buckets(candidates);
 
+    if (logger.is_enabled(logging::log_level::debug)) {
+       int level = 0;
+       for (auto& bucket : buckets) {
+           std::vector<sstring> sstables_generation_and_size;
+           sstables_generation_and_size.resize(bucket.size());
+           std::transform(bucket.begin(),bucket.end(),sstables_generation_and_size.begin(), [](auto& sstable) { return seastar::format("({},{})",sstable->generation(),sstable->data_size()); });
+           logger.debug("size_tiered: level {} (generation,data_size) {}", level, ::join(" ",sstables_generation_and_size));
+           level++;
+       }
+    }
+
     std::vector<sstables::shared_sstable> most_interesting = most_interesting_bucket(std::move(buckets), min_threshold, max_threshold);
     if (most_interesting.empty()) {
         // nothing to do
```

I do not mind that we start a "small compaction" - the questions is should we also have a "large compaction" running in parallel.

**Comment by nyh** [comment]
Thanks. Please don't use the word "level" for what you're printing. It's just a fairly meaningless bucket number, and the parallel compaction decision doesn't use that to make its decision.

As for the last question, no - if we started a small compaction, we would not start a large compaction in parallel - we would wait until the small compaction ends. If we never ever get around to starting a large compaction, then we have a problem. I think we have various heuristics to fix that (if a bucket has many sstables, it is chosen before the smallest bucket). Is that what you're seeing - small sstables being constantly compacted but the bigger sstables never get compacted?

**Comment by slivne** [comment]
Yes we are seeing that files in "bucket 1" (level 1 in the printout) are compacted  and files in "bucket 3" (level 3 in the printout) are not compacted - what is the heuristic ?

**Comment by nyh** [comment]
Caveat lector: @raphaelsc is the expert on this, I'm just writing from memory and brief look at the code.

After we divide the sstables to "buckets" of similarly-sized sstables (how we do that is a separate issue), we have a most_interesting_bucket() function to pick the most interesting bucket for compaction. As far as I can tell, this function currently prefers to compact smaller table first. 

If I understand correctly, Cassandra doesn't pick the smallest tables to compact first, but rather picks the "hottest" bucket - one which contains the sstables which are read more frequently.
 
If I understand you correctly, you're worried about "starvation" of the biggest sstable - we'll keep compacting the smallest sstables and never reach the biggest one. However, I think the theory is that we cannot starve compaction because we dedicate enough resources to it so it never falls behind, and we'll always eventually reach the point where there is **nothing** left to compact. I don't know, however, if this theory holds water in practice in the current code. Do you see that compaction is always running, suggesting it might not be keeping up?

There's another issue to consider: Imagine an unlucky situation where most of the data of each shard sits in one huge sstable (this seems to be the case in your output). We'll never compact this huge sstable until we have four of them (which is our threshold). But at that point, we already have more than four times as much data as we really need! Maybe you can't afford this sort of 4-fold disk waste.

I think Cassandra's solution to this issue is the "tombstone_threshold" parameter: "The ratio of garbage-collectable tombstones to all contained columns. If the ratio exceeds this limit, Cassandra starts compaction on that table alone, to purge the tombstones.". @raphaelsc do we have anything like that? I seem to remember you were talking about such a feature.

**Comment by raphaelsc** [comment]
On Sun, Feb 19, 2017 at 12:19 PM, Roy Dahan <notifications@github.com>
wrote:

> *Installation details*
> Scylla version (or git commit hash): 1.6+
>
> we know there is no compaction level 3 of ~900M files ... why does
> compaction select files in level 1 - why does it not start a compaction in
> level 3 when we have 3 such files ?
>
BTW, size-tiered strategy has no notion of level. I suppose Nadav already
mentioned it. Level in your example is actually bucket id.

Size tiered always look for the bucket with smallest similar-sized sstables
which has at least *** MIN_THRESHOLD *** sstables in it. Min threshold is
usually 4, that's why bucket id == 3 wasn't chosen.

**Comment by raphaelsc** [comment]
On Mon, Feb 20, 2017 at 12:39 PM, Raphael S. Carvalho <
raphaelsc@scylladb.com> wrote:

>
>
> On Sun, Feb 19, 2017 at 12:19 PM, Roy Dahan <notifications@github.com>
> wrote:
>
>> *Installation details*
>> Scylla version (or git commit hash): 1.6+
>>
>> we know there is no compaction level 3 of ~900M files ... why does
>> compaction select files in level 1 - why does it not start a compaction in
>> level 3 when we have 3 such files ?
>>
> BTW, size-tiered strategy has no notion of level. I suppose Nadav already
> mentioned it. Level in your example is actually bucket id.
>
> Size tiered always look for the bucket with smallest similar-sized
> sstables which has at least *** MIN_THRESHOLD *** sstables in it. Min
> threshold is usually 4, that's why bucket id == 3 wasn't chosen.
>


I confused level 3 with 3 sstables, but anyway, level 3 wasn't chosen
because of the first rule I mentioned above. Size tiered always look for
the smallest tiers.

**Comment by raphaelsc** [comment]
On Sun, Feb 19, 2017 at 1:05 PM, Shlomi Livne <notifications@github.com>
wrote:

> the printouts are here
>
> diff --git a/sstables/compaction_strategy.cc b/sstables/compaction_strategy.cc
> index 8d80dd4..0094726 100644
> --- a/sstables/compaction_strategy.cc
> +++ b/sstables/compaction_strategy.cc
> @@ -609,6 +609,17 @@ compaction_descriptor size_tiered_compaction_strategy::get_sstables_for_compacti
>
>      auto buckets = get_buckets(candidates);
>
> +    if (logger.is_enabled(logging::log_level::debug)) {
> +       int level = 0;
> +       for (auto& bucket : buckets) {
> +           std::vector<sstring> sstables_generation_and_size;
> +           sstables_generation_and_size.resize(bucket.size());
> +           std::transform(bucket.begin(),bucket.end(),sstables_generation_and_size.begin(), [](auto& sstable) { return seastar::format("({},{})",sstable->generation(),sstable->data_size()); });
> +           logger.debug("size_tiered: level {} (generation,data_size) {}", level, ::join(" ",sstables_generation_and_size));
> +           level++;
> +       }
> +    }
> +
>      std::vector<sstables::shared_sstable> most_interesting = most_interesting_bucket(std::move(buckets), min_threshold, max_threshold);
>      if (most_interesting.empty()) {
>          // nothing to do
>
> I do not mind that we start a "small compaction" - the questions is should
> we also have a "large compaction" running in parallel.
>

It's true that we only need to start a small compaction in parallel to a
large one, not the other way around. That's what you mean right? I see no
problem with size tiered itself. It's about our parallel compaction
heuristic which lies outside the strategy.

**Comment by raphaelsc** [comment]
On Sun, Feb 19, 2017 at 7:58 PM, nyh <notifications@github.com> wrote:

> Caveat lector: @raphaelsc <https://github.com/raphaelsc> is the expert on
> this, I'm just writing from memory and brief look at the code.
>
> After we divide the sstables to "buckets" of similarly-sized sstables (how
> we do that is a separate issue), we have a most_interesting_bucket()
> function to pick the most interesting bucket for compaction. As far as I
> can tell, this function currently prefers to compact smaller table first.
>
> If I understand correctly, Cassandra doesn't pick the smallest tables to
> compact first, but rather picks the "hottest" bucket - one which contains
> the sstables which are read more frequently.
>
> If I understand you correctly, you're worried about "starvation" of the
> biggest sstable - we'll keep compacting the smallest sstables and never
> reach the biggest one. However, I think the theory is that we cannot starve
> compaction because we dedicate enough resources to it so it never falls
> behind, and we'll always eventually reach the point where there is
> *nothing* left to compact. I don't know, however, if this theory holds
> water in practice in the current code. Do you see that compaction is always
> running, suggesting it might not be keeping up?
>
Yes, strategy should look at small tiers first because it would fall behind
otherwise in number of sstables, which would result in degradation of read
perf.
I was thinking that parallel compaction may actually change the game here.
It's possible that we could compact all tiers in parallel. So compaction
manager should be changed to only allow compaction of smaller weight for a
given column family in parallel.

BTW, I was also thinking that we could increase WEIGHT_LOG_BASE to
something like 10.


> There's another issue to consider: Imagine an unlucky situation where most
> of the data of each shard sits in one huge sstable (this seems to be the
> case in your output). We'll never compact this huge sstable until we have
> four of them (which is our threshold). But at that point, we already have
> more than four times as much data as we really need! Maybe you can't afford
> this sort of 4-fold disk waste.
>
> I think Cassandra's solution to this issue is the "tombstone_threshold"
> parameter: "The ratio of garbage-collectable tombstones to all contained
> columns. If the ratio exceeds this limit, Cassandra starts compaction on
> that table alone, to purge the tombstones.". @raphaelsc
> <https://github.com/raphaelsc> do we have anything like that? I seem to
> remember you were talking about such a feature.
>

I actually wrote a patchset for it, but I had to move to something else,
look:
[RFC 0/9] introduce tombstone removal compaction


We indeed need this tombstone compaction for such scenarios. My plan is to
work on it ASAP.

**Comment by nyh** [comment]
On Mon, Feb 20, 2017 at 6:07 PM, Raphael Carvalho <notifications@github.com>
wrote:

>
> >
> > I think Cassandra's solution to this issue is the "tombstone_threshold"
> > parameter: "The ratio of garbage-collectable tombstones to all contained
> > columns. If the ratio exceeds this limit, Cassandra starts compaction on
> > that table alone, to purge the tombstones.". @raphaelsc
> > <https://github.com/raphaelsc> do we have anything like that? I seem to
> > remember you were talking about such a feature.
> >
>
> I actually wrote a patchset for it, but I had to move to something else,
> look:
> [RFC 0/9] introduce tombstone removal compaction
>
> We indeed need this tombstone compaction for such scenarios. My plan is to
> work on it ASAP.
>

Shlomi corrected me in that in his use-case, there are no tombstones -
rather, the data is being *overwritten*. This is harder to figure out,
although with some effort we can theoretically do this (perhaps using the
intersection estimator that I'm not sure we ever implemented). I don't
think Cassandra has such a feature, though?

**Comment by slivne** [comment]
@raphaelsc 

> I actually wrote a patchset for it, but I had to move to something else,
> look:
> [RFC 0/9] introduce tombstone removal compaction
> 
> 
> We indeed need this tombstone compaction for such scenarios. My plan is to
> work on it ASAP.
> 

Is this in 

if so I'll close this for now

**Comment by raphaelsc** [comment]
On Sun, Jan 21, 2018 at 9:40 AM, Shlomi Livne <notifications@github.com>
wrote:

> @raphaelsc <https://github.com/raphaelsc>
>
> I actually wrote a patchset for it, but I had to move to something else,
> look:
> [RFC 0/9] introduce tombstone removal compaction
>
> We indeed need this tombstone compaction for such scenarios. My plan is to
> work on it ASAP.
>
> Is this in
>
> if so I'll close this for now
>

it is.

thanks.
